server/resp/src/main/java/org/infinispan/server/resp/commands/list/blocking/BLPOP.java

```diff
@@ -0,0 +1,230 @@
+         // Listener can lose events during its install, so we need to poll again
+         // and if we get values complete the listener future. In case of exception
+         // completeExceptionally
+         pollAllKeys(listMultimap, filterKeys).whenComplete((v, t2) -> {
+            // If second poll fails, remove listener and complete exceptionally
+            if (t2 != null) {
+               pubSubListener.completeExceptionally(t);
+            }
+            // Complete poll stage
+            pubSubListener.completePoll(v);
+            if (v == null) {
+               // If no value, start a timer if required
+               pubSubListener.startTimer(timeout);
+            }
+         });
+      });
+      return pubSubListener.getFuture();
+   }
+
+   private CompletionStage<Collection<byte[]>> pollAllKeys(EmbeddedMultimapListCache<byte[], byte[]> listMultimap,
+         List<byte[]> filterKeys) {
+      var pollStage = pollKeyValue(listMultimap, filterKeys.get(0));
+      for (int i = 1; i < filterKeys.size(); ++i) {
+         var keyChannel = filterKeys.get(i);
+         pollStage = pollStage.thenCompose(
+               v -> (v == null || v.isEmpty())
+                     ? pollKeyValue(listMultimap, keyChannel)
+                     : CompletableFuture.completedFuture(v));
+      }
+      return pollStage;
+   }
+
+   @Listener(clustered = true)
+   public static class PubSubListener {
+      private final Channel channel;
+      EmbeddedMultimapListCache<byte[], byte[]> multimapList;
+      public AdvancedCache<Object, Object> cache;
+      ScheduledFuture<?> scheduledTimer;
+      private boolean listenerAdded = false;
+      private CompletableFuture<Collection<byte[]>> listnerFuture = new CompletableFuture<>();
+      private CompletableFuture<Collection<byte[]>> pollFuture = new CompletableFuture<>();
+      private CompletableFuture<Collection<byte[]>> future;
+      Resp3Handler handler;
+
+      public void setListenerAdded(boolean listenerAdded) {
+         this.listenerAdded = listenerAdded;
+      }
+
+      public PubSubListener(Channel channel, Resp3Handler handler, AdvancedCache<Object, Object> cache,
+            EmbeddedMultimapListCache<byte[], byte[]> mml) {
+         this.channel = channel;
+         this.multimapList = mml;
+         this.cache = cache;
+         this.handler = handler;
+         // This future sync poll thread and listener. It waits for the poll stage completion
+         // and its value returned if available otherwise wait for events
+         future = pollFuture.thenCompose((v) -> v != null ? CompletableFuture.completedFuture(v) : listnerFuture)
+               .whenComplete((ignore_v, ignore_t) -> {
+                  try {
+                     this.deleteTimer();
+                     if (listenerAdded) {
+                        cache.removeListenerAsync(this);
+                     }
+                  } catch (Exception ex) {
+                     System.out.println(ex.toString());
```

Comment:
Do we want to catch all Exception here? I would say just let the error propagate to the Stage instead.